libs-scala/oracle-testing/src/main/scala/testing/oracle/OracleAround.scala

```diff
@@ -5,86 +5,127 @@ package com.daml.testing.oracle
+
+  def createNewUniqueRandomUser(): RichOracleUser = createRichOracleUser { stmt =>
+    val id = Random.nextInt(1000 * 1000) + 1
+    val user = OracleUser(s"U$id", id)
+    createUser(stmt, user.name, user.pwd)
+    logger.info(s"New unique random Oracle user created $user")
+    user
   }
 
-  protected def createNewUser(name: String, pwd: String = "hunter2"): User = {
-    Using.Manager { use =>
-      val con = use(
-        DriverManager.getConnection(
-          s"jdbc:oracle:thin:@localhost:$port/ORCLPDB1",
-          "sys as sysdba", // TODO this is needed for being able to grant the execute access for the sys.dbms_lock below. Consider making this configurable
-          systemPwd,
-        )
-      )
-      if (!checkUserExists(con, name)) {
-        val stmt = con.createStatement()
-        stmt.execute(s"""create user $name identified by $pwd""")
-        stmt.execute(s"""grant connect, resource to $name""")
-        stmt.execute(
-          s"""grant create table, create materialized view, create view, create procedure, create sequence, create type to $name"""
-        )
-        stmt.execute(s"""alter user $name quota unlimited on users""")
-
-        // for DBMS_LOCK access
-        stmt.execute(s"""GRANT EXECUTE ON SYS.DBMS_LOCK TO $name""")
-        stmt.execute(s"""GRANT SELECT ON V_$$MYSTAT TO $name""")
-        stmt.execute(s"""GRANT SELECT ON V_$$LOCK TO $name""")
-      } else false
-    }.get
-    User(name, pwd)
+  def createOrReuseUser(name: String): RichOracleUser = createRichOracleUser { stmt =>
+    val user = OracleUser(name.toUpperCase, -1)
+    if (!userExists(stmt, user.name)) {
+      createUser(stmt, user.name, user.pwd)
+      logger.info(s"User $name not found: new Oracle user created $user")
+    } else {
+      logger.info(s"User $name already created: re-using existing Oracle user $user")
+    }
+    user
   }
 
-  protected def dropUser(name: String): Unit = {
-    Using.Manager { use =>
-      val con = use(
+  private def userExists(stmt: Statement, name: String): Boolean = {
+    val res = stmt.executeQuery(
+      s"""SELECT count(*) AS user_count FROM all_users WHERE username='$name'"""
+    )
+    res.next()
+    res.getInt("user_count") > 0
+  }
+
+  private def createUser(stmt: Statement, name: String, pwd: String): Unit = {
+    stmt.execute(s"""create user $name identified by $pwd""")
+    stmt.execute(s"""grant connect, resource to $name""")
+    stmt.execute(
+      s"""grant create table, create materialized view, create view, create procedure, create sequence, create type to $name"""
+    )
+    stmt.execute(s"""alter user $name quota unlimited on users""")
+
+    // for DBMS_LOCK access
+    stmt.execute(s"""GRANT EXECUTE ON SYS.DBMS_LOCK TO $name""")
+    stmt.execute(s"""GRANT SELECT ON V_$$MYSTAT TO $name""")
+    stmt.execute(s"""GRANT SELECT ON V_$$LOCK TO $name""")
+    ()
+  }
+
+  private def createRichOracleUser(createBody: Statement => OracleUser): RichOracleUser = {
```

Comment:
Shouldn't this use transactions to roll back the effect of failed `createBody` calls?

Reply:
I was thinking it wont be needed because usually the first one fails, and if from the rest is something failing it wont work anyway...but makes sense making it more robust already...doing it :+1: